src/masterless.py

```diff
@@ -80,11 +60,62 @@ def set_versions(stackname, *repolist):
 
         arglist.append((repo, known_formula_map[repo], rev))
 
-    if not arglist:
+    # test given revisions actually exist in formulas
+    for name, _, revision in arglist:
+        path = join(config.PROJECT_PATH, "cloned-projects", name)
+        if not os.path.exists(path):
+            LOG.warn("couldn't find formula %r locally, revision check skipped", path)
+            continue
+
+        with lcd(path), settings(warn_only=True):
+            ensure(local("git fetch --quiet").succeeded, "failed to fetch remote refs for %s" % path)
+            ensure(local("git cat-file -e %s^{commit}" % revision).succeeded, "failed to find ref %r in %s" % (revision, name))
+
+    return arglist
+
+#
+#
+
+@task
+@requires_project
+@requires_master_server_access
+def launch(pname, instance_id=None, alt_config='standalone', *repolist):
+    pdata = project.project_data(pname)
+    # ensure given alt config has masterless=True
+    ensure(pdata['aws-alt'], "project has no alternate configurations")
+    ensure(alt_config in pdata['aws-alt'], "unknown alt-config %r" % alt_config)
+    ensure(pdata['aws-alt'][alt_config]['ec2']['masterless'], "alternative configuration %r has masterless=False" % alt_config)
+    repolist = parse_validate_repolist(pdata, *repolist)
+
+    import cfn
+    cfn.launch(pname, instance_id, alt_config, formula_revisions=repolist)
+
+
+@task
+@requires_aws_stack
+@requires_master_server_access
+@requires_masterless
+def update(stackname):
+    import cfn
```

Comment:
can it be imported at the top level in this file? Some clashing names?

Reply:
I kind of trusted my previous self to have made the right decision, but I don't recall a circular dependency ever occurring in masterless.py. I generally don't like importing from other task modules because there is no clear hierarchy, but that itself is just avoiding a smell

Reply:
ah, it wasn't a circular dependency, it's that if you define a task in one module and then import that module into another, then that other module now has the tasks `secondmodule.firstmodule` task in it:

```
    masterless.cfn.aws_delete_stack      tell aws to...
    masterless.cfn.aws_launch_instance
    masterless.cfn.aws_stack_list        returns a l...
    masterless.cfn.aws_update_stack      Updates the...
    masterless.cfn.cmd
```